src/transformers/trainer.py

```diff
@@ -1450,15 +1450,23 @@ def patched_optimizer_step(optimizer, barrier=False, optimizer_args={}):
 
     def _xla_sharded_dataloader(self, dataloader):
         if is_torch_tpu_available():
+            import torch_xla.experimental.xla_sharding as xs
+            import torch_xla.runtime as xr
+            import torch_xla.distributed.parallel_loader as pl
+            num_devices = xr.global_device_count()
+            device_ids = np.arange(num_devices)
+
             sharding_spec = None
             if self.args.spmd_batch_sharding:
-                import torch_xla.experimental.xla_sharding as xs
-                import torch_xla.runtime as xr
-                import torch_xla.distributed.parallel_loader as pl
-                num_devices = xr.global_device_count()
-                device_ids = np.arange(num_devices)
                 mesh = xs.Mesh(device_ids, (num_devices, 1))
                 sharding_spec = xs.ShardingSpec(mesh, (0, 1))
+            elif self.args.spmd_tensor_sharding > 0:
```

Comment:
Specifying `batch` or `fsdp` sharding will silently override the tensor parallelism, can we assert that these flags are exclusive since `tensor_sharding` implies FSDP/batch sharding?

Reply:
Actually, it's intended as you can do tensor_sharding on weights and batch sharding on the input.

Reply:
What I mean is that you can specify:
`--spmd_batch_sharding --spmd_tensor_sharding 4`
but not
`--spmd_fsdp_sharding --spmd_tensor_sharding 4`

Do you think that's clear? If not, I can do a follow up.

Reply:
Got it, thanks Jiewen! I think what we have now is fine. We can follow up later to make the sharding more standard like MaxText has with `ici_*_parallelism` and `dcn_*_parallelism` parameters.

Reply:
Yea, for sure. Does HybridMesh does anything for you in a single slice?

Reply:
Yeah, it will rearrange the tiling assignment to optimize for the ICI connections. I would say we should always use HybridMesh, even for single slice.

Reply:
Trying to see if you have the MFU numbers to compare. I will make a change later.

Reply:
I'll do a quick test on v4-8 to get the MFU difference

Reply:
No worries. It's not a priority.